Add public API to give developers the possibility to adjust the global CSP defaults

LukasReschke opened the pull request:
Allows to inject something into the default content policy. This is for
example useful when you're injecting Javascript code into a view belonging
to another controller and cannot modify its Content-Security-Policy itself.
Note that the adjustment is only applied to applications that use AppFramework
controllers. So for examples the files app and the activity app but not the admin menu.

To use this from your `app.php` use `\OC::$server->getContentSecurityPolicyManager()->addDefaultPolicy($policy)`,
$policy has to be of type `\OCP\AppFramework\Http\EmptyContentSecurityPolicy`.

To test this add something like the following into an `app.php` of any enabled app:

``` php
$manager = \OC::$server->getContentSecurityPolicyManager();
$policy = new \OCP\AppFramework\Http\EmptyContentSecurityPolicy();
$policy->addAllowedFrameDomain('asdf');
$policy->addAllowedScriptDomain('yolo.com');

$policy->allowInlineScript(false);
$manager->addDefaultPolicy($policy);
$policy = new \OCP\AppFramework\Http\EmptyContentSecurityPolicy();
$policy->addAllowedFontDomain('yolo.com');
$manager->addDefaultPolicy($policy);

$policy = new \OCP\AppFramework\Http\EmptyContentSecurityPolicy();
$policy->addAllowedFrameDomain('banana.com');
$manager->addDefaultPolicy($policy);
```

If you now open the files app the policy should be:

``` http
Content-Security-Policy:default-src 'none';script-src 'self' yolo.com 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self' yolo.com;connect-src 'self';media-src 'self';frame-src asdf banana.com 'self'
```

<hr/>

Fixes https://github.com/owncloud/core/issues/16927

@sualko Please test and review so we can get this into 9.0. Especially whether the API is sufficient for your use-case, but from what I understood, it probably does.
@BernhardPosselt Any thoughts?

LukasReschke committed:
Add public API to give developers the possibility to adjust the global CSP defaults

Allows to inject something into the default content policy. This is for
example useful when you're injecting Javascript code into a view belonging
to another controller and cannot modify its Content-Security-Policy itself.
Note that the adjustment is only applied to applications that use AppFramework
controllers.

To use this from your `app.php` use `\OC::$server->getContentSecurityPolicyManager()->addDefaultPolicy($policy)`,
$policy has to be of type `\OCP\AppFramework\Http\ContentSecurityPolicy`.

To test this add something like the following into an `app.php` of any enabled app:
```
$manager = \OC::$server->getContentSecurityPolicyManager();
$policy = new \OCP\AppFramework\Http\ContentSecurityPolicy(false);
$policy->addAllowedFrameDomain('asdf');
$policy->addAllowedScriptDomain('yolo.com');

$policy->allowInlineScript(false);
$manager->addDefaultPolicy($policy);
$policy = new \OCP\AppFramework\Http\ContentSecurityPolicy(false);
$policy->addAllowedFontDomain('yolo.com');
$manager->addDefaultPolicy($policy);

$policy = new \OCP\AppFramework\Http\ContentSecurityPolicy(false);
$policy->addAllowedFrameDomain('banana.com');
$manager->addDefaultPolicy($policy);
```

If you now open the files app the policy should be:

```
Content-Security-Policy:default-src 'none';script-src yolo.com 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src yolo.com 'self';connect-src 'self';media-src 'self';frame-src asdf banana.com 'self'
```

LukasReschke commented:
@PVince81 What we talked about today :smile:

BernhardPosselt commented:
Why false as CSP constructor argument?

BernhardPosselt commented:
Ah, I see this is very hacky. IMHO just subclass CSP again and set the defaults attributes there so you will have a default CSP object and a bare one (sorry phone, hope it's clear)

LukasReschke commented:
> Ah, I see this is very hacky. IMHO just subclass CSP again and set the defaults attributes there so you will have a default CSP object and a bare one (sorry phone, hope it's clear)

Good point. Considering that we can't change the CSP class behaviour as it is public API this probably means we will have `ContentSecurityPolicy` which has default whitelisting and is a subclass of a new `BareContentSecurityPolicy`. Seems doable, though not super happy with the class name :smile:

BernhardPosselt commented:
Empty instead of bare maybe?

BernhardPosselt commented:
Other stuff looks good BTW

LukasReschke commented:
I adjusted the Pull Request. We now have `\OCP\AppFramework\Http\EmptyContentSecurityPolicy` and it's subclasses. Furthermore I fixed a small bug that the EmptyContentSecurityPolicy unit tests discovered. (only whitelisting `unsafe-inline` and forbidding external resources did not work)

This makes the code indeed look cleaner. I'd love to not have the old class called `ContentSecurityPolicy` though because now the subclass sounds like it would be the top-class. But I guess nothing that we can easily change considering that it is part of our public API.

Most changes are unit-tests and so. Just before people get shocked by the size :smile: 

Thanks, @BernhardPosselt.

sualko commented:
Exactly what we need. I did a quick test and it seems to work. I will try to implement and test it in detail tomorrow evening. Thank you Lukas for your effort.

sualko commented:
Works as expected :+1:

LukasReschke commented:
Thanks for the test, @sualko.

@jancborchardt Second review? :speak_no_evil:

jancborchardt commented:
@BernhardPosselt is this a +1 from you? ;)

MorrisJobke commented:
:+1:
